Gracefully shutdown by draining connections

blatyo opened the pull request:
This is hopefully the first step in correctly handling https://github.com/elixir-plug/plug/issues/732

I'm a little lost on when shutdown actually gets called. Wasn't able to find it in plug or phoenix.

blatyo commented:
The failing test seems to be related to a cert being expired.

josevalim commented:
The shutdown is never really called in practice. The best option to
implement this is to have a GenServer that traps exits right after Cowboy.
Then on terminate you drain it.
-- 


*José Valim*
www.plataformatec.com.br
Skype: jv.ptec
Founder and Director of R&D

josevalim commented:
The reply above makes me then wonder if we should really do anything about
it here then. Maybe this should be solved with documentation?
-- 


*José Valim*
www.plataformatec.com.br
Skype: jv.ptec
Founder and Director of R&D

blatyo commented:
Wouldn't the documentation be to trap the exit and then call this?

josevalim commented:
@blatyo yes exactly. Maybe Phoenix could include the drainer by default too but I am not sure how to fit this within Plug.Cowboy itself. I guess we could wrap the cowboy tree in another supervision tree with the drainer in there but that is maybe too much work? Thoughts?

josevalim commented:
Another option is to ship with a `{Plug.Cowboy.Drainer, opts}`  with a process pretty much close to this:

https://github.com/elixir-plug/plug/issues/732#issuecomment-403557433

Then we can do things like:

    {Plug.Cowboy.Drainer, timeout: 5000}

We could also make it drain everything by default. All you need to do is to add it to your sup tree (and phoenix could do it by default if they want to). Thoughts?

whatyouhide commented:
@josevalim really like the idea of providing a drainer out of the box. Not sure about the default behaviour though.

blatyo commented:
While doing it in phoenix would certainly suit my needs, I would argue this should be the default behavior for anyone using cowboy (if they wanted to replicate existing behavior, they could use timeout 0). So, I'd personally go for wrapping in a supervisor. With that approach, you'd pass the timeout as an option when starting the supervisor.

josevalim commented:
Let’s try the supervisor route then!
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

blatyo commented:
I'm plugging away at it

blatyo commented on lib/plug/cowboy/supervisor.ex:
```diff
@@ -0,0 +1,47 @@
+      {Plug.Cowboy.Drainer, opts}
+    ]
+
+    Supervisor.init(children, strategy: :rest_for_one)
```
Do I need to do something to support Elixir <= 1.4? I saw in the tests that it was checking if `Supervisor.start_child/2` was supported.

blatyo commented:
This should be good for another look

josevalim commented on lib/plug/cowboy.ex:
```diff
@@ -108,11 +111,29 @@ defmodule Plug.Cowboy do
-  def shutdown(ref) do
-    :cowboy.stop_listener(ref)
+  @spec shutdown(module(), Keyword.t()) :: :ok | {:error, term}
+  def shutdown(ref, opts \\ []) do
```
Hrm... now I see the problem with this approach. When we shutdown, we also want to shutdown the supervisor and the drainer, instead of leaving the supervisor still alive. The only way I can think to implement this is by changing the API so it receives the supervisor PID. :( Thoughts?

blatyo replied:
I was under the impression that `shutdown/1` was intended to be used with `http/3` and `https/3` exclusively, because the listener is being managed by cowboy. Whereas when managing the process yourself, you could just issue a process exit. I can think about how to handle the case where shutdown also is used to stop the supervisor.

josevalim replied:
@blatyo oh, I see what you mean. Yeah, in this case, we can make this clear in the documentation, :+1:.

blatyo commented:
Just pushed a commit that:
* Adds a much more comprehensive test
* Adds `:drain_check_interval` option, for a setting I originally thought was the timeout
* Changed `:drain_timeout` to set `:shutdown` on the Drainer
* Fixed where options were being looked up to look in `:options` aka cowboy_options

blatyo commented:
This should be good for another look.

josevalim commented:
Hi @blatyo!

I assessed your implementation last night and I thought to myself: "well, I guess this could even be part of ranch then!". So I have opened up an issue on Ranch to discuss about making this part of the *their* supervision tree here: https://github.com/ninenines/ranch/issues/247

They bring important points but the most important in my opinion is that you want to shutdown as many references at the same time as possible, to speed up shutdown time. So I believe we shouldn't go the route in this PR and indeed roll with a separate process that receives all of the references to shutdown concurrently.

What do you think?

Thanks for all of the ideas and work!

blatyo commented:
I get the efficiency angle. My concern is that if in order to achieve that efficiency, you require the user to do a manual step that they're unaware of, they won't do it. In that case, it would be better to do something inefficient if the experience is better overall. If the user were concerned with inefficiency, they could do what's suggested and suspend all the listeners at once and then the Drainer would be a no-op.

josevalim commented:
The problem is that this adds a bunch of redundancy. Plug adds a drainer per listener, then Phoenix adds one for all listeners it has (usually http and https), and then the user may add a third one. Given there is an interest and possibility of multiple drainers, having a building block is more inline with what Plug generally provides.

blatyo commented:
Closed in favor of #14
